### PR TITLE
Fixed test name

### PR DESCRIPTION
**Pull request by Andarist**
Because you have issues disabled on this repository, I'd like to ask smth/discuss.

It seems to me that [this](https://github.com/staltz/callbag-concat/blob/b402eab313264e13080d08efad61b3741fa9f870/readme.js#L49) breaks somewhat `concat`'s transparency, because:
- upwards 1 can be sent with data
- upwards 1s dont have to be synchronous, they can be backpressured etc but now `concat` assumes it has to pull when subscribing to new source

This means that `concat`enated pullable sources might behave differently than a single source combining (I dont mean combining in stream semantics) those 2 sources.

Problem is that `concat` doesnt know if the emitted value is the last one (because 2 is separate message), I've tried to "fix" this with keeping `requesting` & `sending` flags and postponing 2nd action till 1st one finishes to detect sync `1`+`2` emits (happening between `requesting = true` & `requesting = false`) etc, but this seriously hurts my brain 😉 and I cannot achieve my expected results.

To sum it up - do you think it's a problem? I might be overthinking this 😄.

I know that callbags is just a protocol for communication and it doesnt assume much about those things, but still would be great to cover such cases in operators/sources if possible.

### CODE REVIEW

**Comment by staltz** [comment]
Hey @Andarist, you talk about 3 issues actually.

1. concat does not support sink sending upwards 1 with data
2. concat assumes it has to pull when subscribing to new source
3. concat doesn't know if the emitted value is the last one

I think (1) is a real issue that we should fix.

(2) is actually correct behavior. The first member source will be subscribed when concat is subscribed. When the first member source has been fully consumed, the same pull (upwards 1) that caused the `2` from the first member source will be used to pull the first downwards `1` from the second member source. It's hard to explain this in words, but easy in a diagram.

photo_2018-07-23_11-58-57

And about (3), I'm not sure I understand your concern. The sink of a concat basically should not know, and not care about knowing what happens between the end of the first source and the beginning of the second source. Or are you talking about something else?

**Comment by Andarist** [comment]
> And about (3), I'm not sure I understand your concern. The sink of a concat basically should not know, and not care about knowing what happens between the end of the first source and the beginning of the second source. Or are you talking about something else?

Sorry for not making it clear - this was not actual concern but rather noticing why (2) is IMHO some issue.

If you also think that (1) is the issue then (2) "has" also to be one. Referring to your posted image - let's add some data to pulls (assume it works now, fixing issue 1). 3rd pull from the sink is not the same as first pull from the concat to 2nd source, because to be the same it would have to carry the same data with it. If we assume only (1) is fixed then 3rd pull's data gets lost - it doesn't ever get to any source (no source has a chance to process it). At the moment it reaches 1st source the source is already completed and when concat starts to pull the second source directly after greet that pull is data-less.

I hope I've explained my concern in a way that it can be understood - if not please ask me to rephrase.

**Comment by Andarist** [comment]
Friendly 🏓 would love to hear your thoughts about this.

**Comment by staltz** [comment]
@Andarist it took me a while to understand what you meant, but I see the problem now, and indeed issues (1) and (2) are interrelated.

To give a more concrete example to motivate the use case for this, assume that the source gives food, and the sink requests a kind of food. So the sink may request a fruit by sending `1,'fruit'` upwards, and the concatenated sources will respond with the requested fruit. In that case, it makes sense to record the last request made, so that if one source ends when I just requested `1,'vegetable'` from it, then I'll repeat `1,'vegetable'` to the next source.

The trick to designing this operator is to design it from the perspective of the sink, not from the perspective of the sources. From the sink's perspective, it doesn't know and shouldn't care about the internal composition of the source (is it using concat or not). If it's possible to concat sources in ways that are not meaningful or valid, then that's a misuse of concat on those sources.

**Comment by Andarist** [comment]
> From the sink's perspective, it doesn't know and shouldn't care about the internal composition of the source (is it using concat or not).

Totally agreed.

>  In that case, it makes sense to record the last request made, so that if one source ends when I just requested 1,'vegetable' from it, then I'll repeat 1,'vegetable' to the next source.

It's imho not only that. Or it's not as easy as "record last request". There is also a problem (2) that `concat` pulls automatically when subscribing to the next source. As mentioned - concatenated sources should be "transparent" from sink's perspective, so let's illustrate the problem with a code

```js
pipe(
  concat(fromIter([1, 2, 3]), fromIter([4, 5, 6])),
  pullWhen(interval(1000)),
  forEach(value => {
    console.log(value)
  })
)
```

Items should be delivered to the sink in 1 second intervals, but at the moment (due to concat auto pulling behaviour) 3rd and 4th item would be delivered "simultaneously".

While I recognize the problem I have no idea how to fix it properly - I've tried to implement fromIter-like crazy "valve"/variable system to track synchronous sink pulls & source completions, but I've failed. If you have an idea how this can be fixed I'd love to see how this can be implemented. 

I probably won't try to fix this any time soon as I'm not using pullables anywhere - I just like looking through callbags implementations and fix problems for every edge case I can find, lots of fun for me 😅This is over my head though 😢

**Comment by staltz** [comment]
```js
pipe(
  concat(fromIter([1, 2, 3]), fromIter([4, 5, 6])),
  pullWhen(interval(1000)),
  forEach(value => {
    console.log(value)
  })
)
```

> Items should be delivered to the sink in 1 second intervals, but at the moment (due to concat auto pulling behaviour) 3rd and 4th item would be delivered "simultaneously".

Are you sure this happens? It seems to me the pull for downward `t=1,d=3` is not the same pull as downwards `t=2`. So it's the pull for downwards `t=2` that will be reused for `t=1,d=4`.

**Comment by Andarist** [comment]
> Are you sure this happens? It seems to me the pull for downward t=1,d=3 is not the same pull as downwards t=2. So it's the pull for downwards t=2 that will be reused for t=1,d=4.

You are right! I was under strong impression that pullable source would emit on its own 2 after emitting "last value". I see now that it was kinda dumb of me, that would make the pullable source a hybrid, because it would "produce" completion without being asked for it. I see it also relates nicely to javascript  iteratior protocol.

So the only actionable thing here is to record last data sent up with a request, right? Should `filter` & similar operators do the same? 🤔

**Comment by staltz** [comment]
Yes! And good point about `filter` too.

**Comment by Andarist** [comment]
I believe this can be merged now (it's just a simple test name fix).
